### ISSUE TITLE
Clarify `registry.batch` in the documentation

### ISSUE THREAD

**Issue by kevin940726** [WordPress/gutenberg#51075]
## What problem does this address?

Since we're now using React 18 in most if not all of the editors. The primary usage of `registry.batch` is no longer for minimizing React re-renders, since React 18 now [always batches state updates](https://github.com/reactwg/react-18/discussions/21).

## What is your proposed solution?

We should update the doc to reflect the actual usage of the API, and when or why we should use it. https://github.com/WordPress/gutenberg/issues/51075#issuecomment-1567803524 by @jsnajdr is a good starting point!

c.c. @adamziel @youknowriad @jsnajdr

**Comment by jsnajdr** [comment]
I think that `registry.batch` continues to be relevant, as it works very differently from `batch` in React Redux. The React Redux `batch` function is merely a reexport of `ReactDOM.unstable_batchedUpdates`. It's true that in React 18 you don't need that, because state updates are batched automatically.

But if you dispatch 10 actions to a Redux store, it will still call all listeners 10 times, these listeners will run their selector functions, and will schedule a state update if needed. It's batched only at the very end, and a lot of unneeded work is still done.

`registry.batch` is batching much earlier: it will call the store listeners only once after all the 10 dispatches are done. React code (`useSelect` hook) will see only one update, it won't even have anything to batch.

In Redux this kind of batching can be implemented in user-space, by creating a special kind of action and a "higher order reducer" that's used as a wrapper:
```js
function batchingReducer( innerReducer ) {
  return ( state, action ) => {
    if ( action.type === 'BATCH' ) {
      return action.actions.reduce( innerReducer, state );
    }
    return innerReducer( state, action );
  }
}
```
The `@wordpress/data` batching works differently, because there are multiple stores. It pauses calling the listeners at the start of the batch, and restarts and flushes them only at the end. But the outcome is quite identical.

**Comment by youknowriad** [comment]
I agree with @jsnajdr registry.batch is very useful 👍 and very different from React's batching. In addition to the performance optimizations it brings (not running selectors), it is also used for things like "creating a single undo/redo for multiple synchronous updates"

**Comment by kevin940726** [comment]
I see! Thanks for the additional context @jsnajdr .

I don't think many people know this though, as the doc also only mentions reduce rerendering:

> To avoid rerendering the components each time we call `dispatch`, we can wrap the sequential dispatch calls in `batch` which will ensure that the components only call selectors and rerender once at the end of the sequence.

Might be worth updating the doc for this?

**Comment by jsnajdr** [comment]
> Might be worth updating the doc for this?

Yes, although the current wording is not wrong, it's just not rigorously precise. Without batching, components will still "call selectors" but with React 18 auto-batching they won't "rerender" any more.

Another benefit of batching is providing for atomic updates. When batching two actions together, nobody will be able to observe the intermediate state between them. That state is visible only internally, by the dispatched actions.

**Comment by kevin940726** [comment]
I think the doc is suggesting implicitly that `batch` is for minimizing re-renders. The usage code below the doc also says something like:

```js
// This will only rerender the components once.
registry.batch( () => {
	registry.dispatch( someStore ).someAction();
	registry.dispatch( someStore ).someOtherAction();
} );
```

(FWIW, React already batches state updates inside event listeners since React 17 I believe, so the example is also a bit outdated 😅.)

It's technically correct but doesn't provide the value for when or why we want to use it. I'll update the issue title and description! 👍
